napari/utils/events/dataclass.py

```diff
@@ -230,6 +230,18 @@ def _type_to_compare(type_) -> Optional[Callable[[Any, Any], bool]]:
 
     if isinstance(type_, _AnnotatedAlias):
         type_ = type_.__origin__
+    if isinstance(type_, _dc.InitVar):
+        type_ = type_.type
+    if (
+        hasattr(type_, "__origin__")
+        and hasattr(type_.__origin__, "_name")
+        and type_.__origin__._name == "Union"
+        and hasattr(type_, "__args__")
+        and isinstance(type_.__args__, tuple)
+        and len(type_.__args__) == 2
+        and isinstance(None, type_.__args__[1])
+    ):
+        type_ = type_.__args__[0]
```

Comment:
```suggestion
    _args = get_args(type_)
    if get_origin(type_) is Union and len(_args) == 2 and isinstance(None, _args[1]):
        type_ = _args[0]
```

(then add `from typing import Union`)

Reply:
can you also add a short docstring to `_type_to_compare`?

Reply:
oh, and add a comment above that line, saying "get main type from Optional[type]" so it's clear what's going on there

Reply:
Could you check now?

Reply:
I do not know `get_origin` and `get_args`. Thanks for pointing it.